[16.0][MIG] attachment_preview: Migration to 16.0

houzefa-abba opened the pull request:
I squashed a few commits according to https://github.com/OCA/maintainer-tools/wiki/Merge-commits-in-pull-requests

This PR has 2 commits: 1 with the actual migration; 1 where I move code around without changing anything. I can squash them if you feel it's better

I made sure this plays well with [`web_responsive`](https://github.com/OCA/web/tree/16.0/web_responsive) & especially [`web_chatter_position`](https://github.com/OCA/web/tree/16.0/web_chatter_position) with its 3 chatter settings (responsive / bottom / sided)

I added a ROADMAP file with possible future improvements

Overall implementation remains the same as in previous versions:
* patch into attachment list to prepare prev/next buttons in the viewer
* patch into attachment cards to observe clicks on new buttons this
  module adds
* fetch attachment extensions
* add an iframe into the DOM tree next to the main form one, display
  viewer inside it
* add preview button within binary fields

Main changes in this migration:
* update imports/exports to proper JS modules - in particular, this
  fixes "service already defined" console messages we also get in 15.0
* rework FormRenderer injector, previous (legacy) one was no longer used
* patch attachment list / cards with new methods in the mail module
* fix preview button inclusion within binary fields
* fix previous/next viewer buttons

## Screenshots from [runboat](https://runboat.odoo-community.org/webui/builds.html?repo=OCA/knowledge&target_branch=16.0)

2 buttons in attachment cards to preview inline & open in new tab:
![Screenshot at 2023-09-29 15-13-28](https://github.com/OCA/knowledge/assets/6347423/3fb2ebd0-df07-43d9-ba29-fae8df07fc5b)

Inline preview of a PDF:
![Screenshot at 2023-09-29 15-14-48](https://github.com/OCA/knowledge/assets/6347423/db719e8f-f999-45bf-94cc-2edfbc795471)

Inline preview of an ODT (which I navigated to using left arrow at the top of the viewer):
![Screenshot at 2023-09-29 15-15-09](https://github.com/OCA/knowledge/assets/6347423/afef4379-173a-46ca-b14d-40a42c893a67)

Binary field with button to open in new tab:
![Screenshot at 2023-09-29 15-15-57](https://github.com/OCA/knowledge/assets/6347423/97e92c1f-8216-4865-99d8-5e3f47607a22)

hbrunn committed:
[ADD] preliminary port of attachment_preview

hbrunn committed:
[IMP] have titles work everywhere

hbrunn committed:
[FIX] don't depend on document, the relevant code still lives in web

hbrunn committed:
[ADD] translations

hbrunn committed:
[UPD] use viewerJS 0.5.4, incorporate new suported file types

hbrunn committed:
[FIX] readme 8.0 style

gurneyalex committed:
Add OCA as author of OCA addons

In order to get visibility on https://www.odoo.com/apps the OCA board has
decided to add the OCA as author of all the addons maintained as part of the
association.

gurneyalex committed:
fix travis (flake8)

sysadminmatmoz committed:
Slovene translations added

hbrunn committed:
[UPD] use viewerjs 0.5.7

sysadminmatmoz committed:
Translations and templates sync

sbidoul committed:
[UPD] prefix versions with 8.0

pedrobaeza committed:
[MIG] Make modules uninstallable

pedrobaeza committed:
[MIG] Rename manifest files

sbidoul committed:
remove obsolete .pot files [ci skip]

tarteo committed:
[MIG] attachment_preview: Migration to 11.0

[FIX] Lint

[FIX] lint and flake

[ADD] tests

[ADD] tests

[ADD] tests

[ADD] Tests

[ADD] Package python-magic

[ADD] Tests

[FIX] Lint

oca-travis committed:
[UPD] Update attachment_preview.pot

astirpe committed:
[FIX] Robustness when attachment not present

tarteo committed:
[IMP] attachment_preview: Split screen

MiquelRForgeFlow committed:
[FIX] pylint/javascript-lint

tarteo committed:
[MIG] attachment_preview: Migration to 12.0

[REM] Old files

[ADD] Magic to travis file

[FIX] Nagivation refresh widget

sysadminmatmoz committed:
Translated using Weblate (Slovenian)

Currently translated at 100.0% (2 of 2 strings)

Translation: knowledge-12.0/knowledge-12.0-attachment_preview
Translate-URL: https://translation.odoo-community.org/projects/knowledge-12-0/knowledge-12-0-attachment_preview/sl/

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: knowledge-12.0/knowledge-12.0-attachment_preview
Translate-URL: https://translation.odoo-community.org/projects/knowledge-12-0/knowledge-12-0-attachment_preview/

khaeusler committed:
fix the visibility of the preview area

OCA-git-bot committed:
attachment_preview 12.0.1.0.1

marcelsavegnago committed:
Added translation using Weblate (Portuguese (Brazil))

hbrunn committed:
[FIX] attachment_preview: start needs to return a promise

hbrunn committed:
[UPD] hbrunn's mail address

OCA-git-bot committed:
attachment_preview 12.0.1.0.2

yavuzavci committed:
Added translation using Weblate (Turkish)

vancouver29 committed:
[IMP] attachment_preview: black, isort, prettier

vancouver29 committed:
[MIG] attachment_preview: Migration to 15.0

vancouver29 committed:
[IMP] attachment_preview: fix _onPreviewAttachment and remove class o_attachment_preview

luis-ron committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (4 of 4 strings)

Translation: knowledge-15.0/knowledge-15.0-attachment_preview
Translate-URL: https://translation.odoo-community.org/projects/knowledge-15-0/knowledge-15-0-attachment_preview/es/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (4 of 4 strings)

Translation: knowledge-15.0/knowledge-15.0-attachment_preview
Translate-URL: https://translation.odoo-community.org/projects/knowledge-15-0/knowledge-15-0-attachment_preview/it/

OCA-git-bot committed:
[UPD] README.rst

houzefa-abba committed:
[MIG] attachment_preview: Migration to 16.0

Overall implementation remains the same as in previous versions:
* patch into attachment list to prepare prev/next buttons in the viewer
* patch into attachment cards to observe clicks on new buttons this
  module adds
* fetch attachment extensions
* add an iframe into the DOM tree next to the main form one, display
  viewer inside it
* add preview button within binary fields

Main changes in this migration:
* update imports/exports to proper JS modules - in particular, this
  fixes "service already defined" console messages we also get in 15.0
* rework FormRenderer injector, previous (legacy) one was no longer used
* patch attachment list / cards with new methods in the mail module
* fix preview button inclusion within binary fields
* fix previous/next viewer buttons

houzefa-abba committed:
[MOV] attachment_preview: Reorganize JS modules

houzefa-abba commented:
Hi this PR is ready, please review

@vancouver29 @HolgerNahrstaedt @holgern I am tagging you as you worked on https://github.com/OCA/knowledge/pull/363 (15.0 mig)

@hbrunn I am tagging you as you bootstrapped this module, in case you're still around

len-foss commented:
Thank you for migrating this!

Despite having no browser or server error, the preview doesn't seem to work in either FF or Chrome for anything but pdf files. 
With other types that should be supported by viewerJS (doc, docx, ...), the preview is just empty, and opening in a new tab it spawns a "not supported filetype" popup. 
I've tested on two device with the same results. Do you have any idea about what could be wrong?

houzefa-abba commented:
> Thank you for migrating this!
> 
> Despite having no browser or server error, the preview doesn't seem to work in either FF or Chrome for anything but pdf files. With other types that should be supported by viewerJS (doc, docx, ...), the preview is just empty, and opening in a new tab it spawns a "not supported filetype" popup. I've tested on two device with the same results. Do you have any idea about what could be wrong?

Hi @len-foss did you test on runboat? My screenshots in this issue are from runboat; the file with "character formatting" / "paragraph formatting" is an ODT file.

I'm using Chromium mainly, but I just tried with Firefox in an incognito window; no problem either.

At the moment, the only supported file extensions are those [in `attachment_preview/static/src/js/utils.esm.js`](https://github.com/OCA/knowledge/pull/434/files#diff-f01f4b5d281a452a5e34c901194425c0810dc7c337e238984aa0dea568f1c808); doc & docx are not included there.

In this PR I did not update the viewerjs lib; we could update it in a separate PR and add new supported filetypes in there.

astirpe commented:
@houzefa-abba I tried in combination with `web_responsive`, there is a glitch: 

![Screenshot from 2023-10-10 08-47-12](https://github.com/OCA/knowledge/assets/7657635/65e63fbf-382a-44f0-a583-85fec092252b)

len-foss commented:
@houzefa-abba oh thanks, indeed it works fine with odt files, runboat or local. I agree doing the update in a separate PR is the proper way to do it, let's just hope it gets through quickly for once.

I confirm the behavior reported by @astirpe, although I wouldn't call it a glitch. It seems you could trivially fix it by setting the z-index of `attachment_preview_widget` to 199 instead of 999 -- the `full-screen-dropdown` has a z-index of 200, so that would go below. The 200 value is also used by the pdf viewer so it seems to confirm that idea.

houzefa-abba commented:
> @houzefa-abba I tried in combination with `web_responsive`, there is a glitch:
> 
> ![Screenshot from 2023-10-10 08-47-12](https://user-images.githubusercontent.com/7657635/273802380-65e63fbf-382a-44f0-a583-85fec092252b.png)

Good catch! Fixed with z-index change suggested by @len-foss.
I went with z-index=100 as most Odoo elements are between 1-10 and other ones such as pop-ups are >=200.

I also used some `if (x) return;` 1-liners which I noticed in your code 🙃

houzefa-abba commented:
I added widget clean-up in [my last change](https://github.com/OCA/knowledge/compare/0a4b34346119a95496fbe2aa7c6481d50d30e9fd..f55a4349ea949bbba335f56287a0ad73e0e91473), otherwise I was getting multiple widgets in the DOM tree when navigating in and out of form views

hbrunn commented:
/ocabot migration attachment_preview

astirpe commented on attachment_preview/static/src/js/web_views/form_renderer.esm.js:
```diff
@@ -0,0 +1,43 @@
+
+    setup() {
+        var res = this._super(...arguments);
+        this.attachmentPreviewWidget = new AttachmentPreviewWidget(this);
```
At the moment we have two different functionalities for the PDF preview on forms:

- standard Odoo PDF preview
- this module (`attachment_preview`)

Would be nice to have an option to disable this attachment preview on forms for this module (`attachment_preview`). This is a requirement for one of our customers.  So the idea is that the admin could decide to configure the option (enable/disable). This way, when the option is disabled, the standard Odoo preview will be work in place of this module (`attachment_preview`).

Disabling this option can be easily achieved by returning `res` just after the `_super(..)`.

```javascript
        var res = this._super(...arguments);
        if (is_attachment_preview_disabled) return res;
```

houzefa-abba replied:
Yea makes perfect sense. To take it even further, I did ponder whether this module could entirely re-use components added by Odoo to implement their preview feature...

I won't be able to get to this until at least next week.

houssine78 replied:
@houzefa-abba did you get time to get to this or you plan to do it in the near future ? thanks for this PR anyway !

houzefa-abba replied:
Hi this is still in my TODO but it has moved further back 😅
Very unlikely I could get back to this PR until end of May

This per-form setting hinges between fix (imposed by odoo16 adding its own preview system) and evolution; perhaps we could add it as a separate evolution PR so this initial PR can get moving as is?

houzefa-abba replied:
Hi I got back onto this topic for a bit

I just rebased on top of latest 16.0 and added a commit to disable Odoo's default attachment sidebar can show up in views with `<div class="o_attachment_preview" />`

However I kept pondering @astirpe's idea above but I can't see a way it fully makes sense:
* **Global setting**: would be useful to disable our viewer in forms in favor of odoo's own viewer when one still wants to use our viewer somewhere else, but it doesn't make much sense to me: such an odoo app would use 2 different viewers based on view type, potentially displaying the same documents in different ways depending on view type.
* **Per-form setting** (on the `<form>` tag): again, I can't see a use case, especially when we assume our viewer is similar (and in parts more powerful) than odoo's base one.

Could we think about it some more before confirming we need such a setting? In a separate PR possibly?

Thanks

houzefa-abba committed:
[IMP] attachment_preview: Disable base Odoo sidebar

Odoo's default attachment sidebar can show up in views with
`<div class="o_attachment_preview" />`.
Our attachment sidebar would show up on top of it, making the base Odoo
sidebar obsolete. ➔ We disable it.
